### ISSUE TITLE
Feature Suggestion: Support Additional Data Annotation Attributes

### ISSUE THREAD

**Issue by nickcox** [NJsonSchema/NJsonSchema#182]
Would you consider adding support for the full set of annotation attributes [supported by JSON.Net Schema](http://www.newtonsoft.com/jsonschema/help/html/GeneratingSchemas.htm)?

From the page:

- [x] RequiredAttribute | Property will be required.
- [x] MinLengthAttribute | minLength of a string or the minItems in an array.
- [x] MaxLengthAttribute | maxLength of a string or the maxItems in an array.
- [x] RegularExpressionAttribute | String pattern.
- [x] RangeAttribute | Number minimum and maximum.
- [ ] StringLengthAttribute | String minLength and maxLength.
- [ ] EnumDataTypeAttribute | Enum type used to generate enum values.
- [ ] DataTypeAttribute | String format depending on the DataType value. Supports Url, Date, Time, DateTime, EmailAddress, PhoneNumber.
- [ ] UrlAttribute | String uri format.
- [ ] PhoneAttribute | String phone format.
- [ ] EmailAddressAttribute | String email format.

**Comment by rsuter** [comment]
Thanks for listing all of available attributes. We should implement these in the JSON Schema generator...

**Comment by nickcox** [comment]
Awesome. I'm happy to have a crack at this if it's not too difficult.

**Comment by rsuter** [comment]
Yes, I think most attributes should be easy to implement...

**Comment by rsuter** [comment]
Implemented StringLengthAttribute in commit https://github.com/NJsonSchema/NJsonSchema/commit/fe0b60ae23c7fc1f242eb7d2a4fac889b85b1ffb

**Comment by DorianGreen** [comment]
Will it be possible to support custom annotations that **Inherit** from the data annotation classes?

**Comment by rsuter** [comment]
Yes, this should work: 

![image](https://cloud.githubusercontent.com/assets/2603405/18205723/d72be87e-7123-11e6-88f5-a3ad790a36be.png)

**Comment by rsuter** [comment]
No, I think currently it does not work because it looks for the exact attribute name not if an attribute inherits from it: 

        private Attribute TryGetAttribute(IEnumerable<Attribute> attributes, string attributeType)
        {
            return attributes.FirstOrDefault(a => a.GetType().FullName == attributeType);
        }

**Comment by DorianGreen** [comment]
shouldn't changing the method to:
```
private Attribute TryGetAttribute(IEnumerable<Attribute> attributes, string attributeType)
{
       return attributes.FirstOrDefault(a => Type.GetType(attributeType, false, true).IsInstanceOfType(a));
}
```
do the trick?

This way by checking for the DataType attribute, you get all it's subclasses for free (Email, Phone, etc...)
This should also allow for custom sublcassing for validation.

**Comment by rsuter** [comment]
We can use this: https://github.com/NSwag/NSwag/blob/master/src/NSwag.CodeGeneration/Infrastructure/ReflectionExtensions.cs

Can you open a new issue for this feature (instead of discussing it here)?

**Comment by rsuter** [comment]
With https://github.com/NJsonSchema/NJsonSchema/issues/189 phone and email should be handled too.. we should add unit tests for them

**Comment by psandgren** [comment]
Hey! Is this supposed to be implemented? I see the tests, but codegeneration from NSwag doesn't really add the EmailAddressAttribute. Where to look? What to fix? How can I help? :)

**Comment by danielniccoli** [comment]
Is there a chance to also include [CustomValidation](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.customvalidationattribute?view=net-7.0) support?
